solr/core/src/java/org/apache/solr/search/combining/CombineQParserPlugin.java

```diff
@@ -0,0 +1,167 @@
+
+public class CombineQParserPlugin extends QParserPlugin {
+  public static final String NAME = "combine";
+
+  @Override
+  public QParser createParser(
+      String qstr, SolrParams localParams, SolrParams params, SolrQueryRequest req) {
+    return new CombineQParser(qstr, localParams, params, req);
+  }
+
+  static class CombineQParser extends QParser {
+
+    private List<String> unparsedQueries;
+    private QueriesCombiner queriesCombiningStrategy;
+    private List<QParser> queryParsers;
+    private List<Query> queries;
+
+    public CombineQParser(
+        String qstr, SolrParams localParams, SolrParams params, SolrQueryRequest req) {
+      super(qstr, localParams, params, req);
+    }
+
+    @Override
+    public Query parse() throws SyntaxError {
+
+      var queriesToCombineKeys = localParams.getParams("keys");
+      if (queriesToCombineKeys == null) {
+        queriesToCombineKeys = params.getParams(CombinerParams.COMBINER_KEYS);
+      }
+
+      unparsedQueries = new ArrayList<>(queriesToCombineKeys.length);
+      queryParsers = new ArrayList<>(queriesToCombineKeys.length);
+      queries = new ArrayList<>(queriesToCombineKeys.length);
+
+      // nocommit blend localParams and params without needless suffix in local
+      var blendParams = SolrParams.wrapDefaults(localParams, params);
+      queriesCombiningStrategy = QueriesCombiner.getImplementation(blendParams);
+
+      String defType = blendParams.get(QueryParsing.DEFTYPE, DEFAULT_QTYPE);
+
+      for (String queryKey : queriesToCombineKeys) {
+        final var unparsedQuery = blendParams.get(queryKey);
+        final var parser = QParser.getParser(unparsedQuery, defType, req);
+        var query = parser.getQuery();
+        if (query == null) { // sad this can happen
+          query = new MatchNoDocsQuery();
+        }
+        unparsedQueries.add(unparsedQuery);
+        queryParsers.add(parser);
+        queries.add(query);
+      }
+
+      return new CombineQuery(queriesCombiningStrategy, queries);
+    }
+
+    @Override
+    public void addDebugInfo(NamedList<Object> dbg) {
+      super.addDebugInfo(dbg);
+
+      NamedList<NamedList<Object>> queriesDebug = new SimpleOrderedMap<>();
+      String[] queryKeys = req.getParams().getParams(CombinerParams.COMBINER_KEYS);
+      for (int i = 0; i < queries.size(); i++) {
+        NamedList<Object> singleQueryDebug = new SimpleOrderedMap<>();
+        singleQueryDebug.add("querystring", unparsedQueries.get(i));
+        singleQueryDebug.add("queryparser", queryParsers.get(i).getClass().getSimpleName());
+        singleQueryDebug.add("parsedquery", QueryParsing.toString(queries.get(i), req.getSchema()));
+        singleQueryDebug.add("parsedquery_toString", queries.get(i).toString());
+        queriesDebug.add(queryKeys[i], singleQueryDebug);
+      }
+      dbg.add("queriesToCombine", queriesDebug);
+    }
+  }
+
+  static final class CombineQuery extends ExtendedQueryBase
```

Comment:
A "SelfExecutingQuery" isn't normal; Lucene isn't going to work with this thing since it doesn't implement getWeight nor can it.  But users don't need to know/care.